### PR TITLE
remove redundant iterator call to data fetcher in loops

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

#9052 moved the iterator management from the fit loop to the training epoch loop.
the dataloader gets called with `iter()` but later on also with `enumerate()` on top of it. this will not play nicely in #8950 where it's important to track the state of an iterator. 

This PR removes the additional `iter()` call.
A full integration test for iterator state is present in #8950 .

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
    NOTNEEDED since the change linked above was only introduced in master recently.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/utilities.py [hunk 1]
```diff
@@ -105,9 +105,11 @@ def _process_training_step_output(
     return results, hiddens
 
 
-def _prepare_dataloader_iter(dataloader_iter: Iterator, batch_idx: int) -> Iterator:
+def _prepare_dataloader_iter(data_fetcher: AbstractDataFetcher, batch_idx: int) -> Iterator:
     """Attach the dataloader"""
-    if not isinstance(dataloader_iter, DataLoaderIterDataFetcher):
-        dataloader_iter = enumerate(dataloader_iter, batch_idx)
-    # restore iteration
+    if not isinstance(data_fetcher, DataLoaderIterDataFetcher):
```
Comment (aazzolini) [inline pytorch_lightning/loops/utilities.py:110]: why not simply move this into the contract of this iterator? basically make the contract such that these data fetchers will always return the batch_idx as part of it.

File: pytorch_lightning/loops/utilities.py [hunk 2]
```diff
@@ -105,9 +105,11 @@ def _process_training_step_output(
     return results, hiddens
 
 
-def _prepare_dataloader_iter(dataloader_iter: Iterator, batch_idx: int) -> Iterator:
+def _prepare_dataloader_iter(data_fetcher: AbstractDataFetcher, batch_idx: int) -> Iterator:
```
Comment (aazzolini) [inline pytorch_lightning/loops/utilities.py:108]: How much do we actually need to expose the "AbstractDataFetcher" here versus just an iterator? It seems to me that we're really not adding much value by exposing a new interface here. 

The custom logic below really should be part of the iterator implementation itself as i comment below.

**Review by aazzolini — COMMENTED** [review]
Making evaluation_loop depend on fetcher doesn't seem needed to me -- do we need anything other than an iterator proper?

File: pytorch_lightning/loops/utilities.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/loops/utilities.py:108]: yes, the iterator type annotation would have sufficed here

File: pytorch_lightning/loops/utilities.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/utilities.py:110]: yes, I suppose could be done like that :)
